### PR TITLE
SAA-875 apply activity end date changes to the associated schedules, allocations and planned deallocations.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
I have tested these changes locally running the UI against them.

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/AllocationTest.kt [hunk 1]
```diff
@@ -300,4 +300,36 @@ class AllocationTest {
       ),
     )
   }
+
```
Comment (iterative-it) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/AllocationTest.kt:303]: I don't think there's a test for the `.filterNot { allocation -> allocation.status(PrisonerStatus.ENDED) }` filter although I could have missed it when looking through the rest of this file. Worth adding one?

Comment (matt-crowson-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/AllocationTest.kt:303]: Done.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivityScheduleTest.kt [hunk 2]
```diff
@@ -925,4 +925,68 @@ class ActivityScheduleTest {
...
+
+    schedule.allocations().forEach { allocation -> assertThat(allocation.endDate).isEqualTo(tomorrow) }
+
+    schedule.endDate = today
+
+    schedule.allocations().forEach { allocation -> assertThat(allocation.endDate).isEqualTo(today) }
+  }
+
+  @Test
+  fun `change to schedule end date is is not applied to ended allocations`() {
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivityScheduleTest.kt:960]: @iterative-it

Missing test added here.

**Review by iterative-it — APPROVED** [review]
LGTM

**Pull request merged**